### PR TITLE
[UNR-1197] Fix: Call query delegate even on query failure 

### PR DESCRIPTION
**Pull request by tenevdev**
#### Description
Query response delegates are now called regardless of the response status code. The previous behaviour didn't allow handlers to implement custom retry logic in case of query failure.

#### Release note
Bugfix: Execute query response delegates even if response status shows failure. This allows handlers to implement custom retry logic.

#### Tests
Added unit tests in a newly created module.

#### Documentation
release note

#### Primary reviewers
@firtina-improbable @joshuahuburn

### CODE REVIEW

**Comment by danielimprobable** [comment]
@tenevdev could you flag that you are adding a test with the GDK, I suspect that the GDK team would have opinions on this.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -1195,22 +1195,20 @@ void USpatialReceiver::OnCreateEntityResponse(Worker_CreateEntityResponseOp& Op)
...
-			RequestDelegate->ExecuteIfBound(Op);
-		}
-		else
-		{
-			UE_LOG(LogSpatialReceiver, Warning, TEXT("Recieved EntityQueryResponse but with no delegate set, request id: %d, number of entities: %d, message: %s"), Op.request_id, Op.result_count, UTF8_TO_TCHAR(Op.message));
-		}
+		UE_LOG(LogSpatialReceiver, Error, TEXT("EntityQuery failed: request id: %d, message: %s"), Op.request_id, UTF8_TO_TCHAR(Op.message));
+	}
+
+	auto RequestDelegate = EntityQueryDelegates.Find(Op.request_id);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: No auto pls

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Thanks, I'll fix this, but it was already like this.

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Misc/AutomationTest.h"
+#include "Interop/SpatialReceiver.h"
+
+namespace SpatialReceiverTestUtils {
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]: Add newline before {

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	bool SpatialReceiverOnEntityQueryResponseTest(Worker_EntityQueryResponseOp& TestOp)
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]: Can the TestOp be const reference?

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	{
+		USpatialReceiver* Receiver = NewObject<USpatialReceiver>();
+		Worker_RequestId RequestId = 0;
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]:
```suggestion
		const Worker_RequestId RequestId = 0;
```

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+IMPLEMENT_SIMPLE_AUTOMATION_TEST(FSpatialReceiverOnEntityQueryResponseTestStatusCodeSuccess, "SpatialGDK.USpatialReceiver.OnEntityQueryResponse.CallsDelegateIfStatusCodeIsSuccess", EAutomationTestFlags::EditorContext | EAutomationTestFlags::EngineFilter)
+
+bool FSpatialReceiverOnEntityQueryResponseTestStatusCodeSuccess::RunTest(const FString& Parameters)
+{
+	Worker_EntityQueryResponseOp TestQueryResponseOp{};
+	TestQueryResponseOp.status_code = Worker_StatusCode::WORKER_STATUS_CODE_SUCCESS;
+
+	return SpatialReceiverTestUtils::SpatialReceiverOnEntityQueryResponseTest(TestQueryResponseOp);
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]: What is the error message it this function fails?

File: SpatialGDK/Source/SpatialGDKTest/Public/SpatialGDKTestModule.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "Developer/Settings/Public/ISettingsContainer.h"
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDKTest/Public/SpatialGDKTestModule.h]: Why do we need these headers?

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp (hunk 5 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]: It shows red in the editor with a link to this line. A developer would navigate to this and understand the delegate isn't being called. They will then put some breakpoints and investigate why.

When I have some learning time, I'll try to make some improvements to the hot-reloading flow to make running tests while making changes more intuitive. Currently it's a bit of a pain so I wouldn't encourage people to write tests yet.

File: SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp (hunk 3 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDKTest/Private/Tests/SpatialReceiverTest.cpp]: Yes.

File: SpatialGDK/Source/SpatialGDKTest/Public/SpatialGDKTestModule.h (hunk 6 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDKTest/Public/SpatialGDKTestModule.h]: Copy-paste error. Removed now.

**Pull request merged**